guidance/library/_gen.py

```diff
@@ -98,107 +92,94 @@ def gen(
-    # if name is not None and not list_append:
-    #     lm[name] = ""
+        stop_regex += [quote_regex(s) for s in stop]
+        if len(stop_regex) == 1:
+            gen_stop = stop_regex[0]
+        else:
+            gen_stop = "|".join("(" + s + ")" for s in stop_regex)
 
-    # define the generation pattern
-    if regex is not None:
-        pattern = regex_grammar(regex)
-    else:
-        pattern = zero_or_more(any_char())
+    if regex is None:
+        regex = r"(?s:.*)"
+    if save_stop_text is True:
+        save_stop_text = str(name) + "_stop_text"
+    if not isinstance(save_stop_text, str):
+        save_stop_text = None
 
     tagged_name = "__LIST_APPEND:" + name if list_append and name is not None else name
 
-    # define any capture group for non-tool calls
-    if name is not None and tools is None:
-        pattern = capture(pattern, name=tagged_name)
-
-    # limit the number of tokens
-    pattern = token_limit(pattern, max_tokens)
-
-    # define the stop pattern
-    if stop is False or len(stop + stop_regex) == 0:
-        stop_pattern = ""
-    else:
-        stop_pattern = select(stop + stop_regex)
-        if save_stop_text is True:
-            save_stop_text = str(name) + "_stop_text"
-        if isinstance(save_stop_text, str):
-            stop_pattern = capture(stop_pattern, name=save_stop_text)
-        stop_pattern = commit_point(stop_pattern, hidden=True)
-
-    # single generation
-    start_pos = len(str(lm))
     if tools is not None:
-        with block(tagged_name):
-            tools = [Tool(callable=x) if not isinstance(x, Tool) else x for x in tools]
-            init_token_count = lm.token_count
-            gen_grammar = pattern + select(
-                [stop_pattern]
-                + [
-                    capture(
-                        commit_point(x.call_grammar, hidden=hide_tool_call),
-                        name=f"tool{i}",
-                    )
-                    for i, x in enumerate(tools)
-                ]
+        tools = [Tool(callable=x) if not isinstance(x, Tool) else x for x in tools]
+        options = [
+            Gen(body_regex=regex, stop_regex=gen_stop, save_stop_text=save_stop_text, max_tokens=max_tokens)
+        ]
+        for i, tool in enumerate(tools):
+            # Infer a regex that will match the start of a tool call
+            tool_call_prefix = tool.call_grammar.forced_prefix()
+            if len(tool_call_prefix) < 4:
+                # TODO: alternatively check that the prefix contains the name (case insensitive) of the tool?
+                # anything shorter is probably far too ambiguous
+                raise ValueError(f"Could not infer unambiguous tool call prefix for tool {tool.name}")
+            options.append(
+                capture(
+                    Gen(body_regex=regex, stop_regex=quote_regex(tool_call_prefix), max_tokens=max_tokens),
+                    name=f"tool{i}"
+                )
             )
```

Comment:
Admittedly a bit of a hack. We don't currently support stopping a gen on an arbitrary context-free grammar (or even a RegularGrammar for that matter), we need a bona-fide regex. Here, I'm introspecting on the tool's `call_grammar` to see if it has a forced prefix (e.g. `"calc("`) and using that prefix as the stop sequence.

If the prefix is very short, I'm raising an exception just so that we're not accidentally forcing tool calls all over the place if the prefix isn't very unique/identifiable. This is kind of arbitrary.

I rather like the approach of forcing the rest of the tool call's grammar after hitting its prefix. This way we actually guarantee correct call syntax rather than *hoping* the model gets it right and actually calling the tool if and only if it does.

We could alternatively explicitly use something like the tool's name followed by an open paren as the stop sequence, but this may not be the actual call grammar the user expects.

@Harsha-Nori would appreciate your eyes on this (even if this is a "temporary" implementation)

Reply:
The LLG API allows any regular expression as stop - it be the regular grammar thingy, doesn't have to be string.

Reply:
Ah, my mistake. I could try and turn the grammar into a regex here then (maybe falling back to the forced prefix trick if that fails), but I actually kind of like that this approach actually forces using the grammar once it's "sufficiently convinced" that it's the one you want... I suppose the other problem is that our regex don't support captures at the moment (is that something you'd be open to implementing at some point?)

Reply:
I like the forcing grammar once you see "calc(" as well!

Capture groups with derivative regexes are an open research question...